### DISCUSSION TITLE
greater/deeper use of Rx & observables

### DISCUSSION THREAD

**Discussion by BenMakesGames** [AvaloniaUI/Avalonia#6312] [Ideas]
**context:** I'm an experienced Angular developer, and now find myself starting a desktop application at a new job. the other devs here are learning WPF, so I've started doing the same, but found WPF very... verbose compared to web frameworks (the Blazor dev here shares that opinion). AvaloniaUI, however, appears to solve a lot of these problems; I've shared AvaloniaUI with the other devs, but I'm still in the process of learning it, and am not (yet) at a point where I feel confident saying "this is what we gotta' use".

**suggestion:** I'd love to see greater, low-level support for observables in AvaloniaUI. I think AvaloniaUI's addition of binding to observables in the view is incredible! https://docs.avaloniaui.net/docs/data-binding/binding-to-tasks-and-observables and I suspect a lot of WPF/XAML's change notification & dependency property boilerplate could be done away with if observables were used more-thoroughly.

**philosophical question:** how much is the inclusion of observables and ReactiveUI seen as a draw by experienced AvaloniaUI developers? (I've noticed some mixed feelings about the inclusion of Splat, for example https://github.com/AvaloniaUI/Avalonia/discussions/6157 ; I could imagine there being mixed feelings about ReactiveUI.) observables are a core part of Angular; is that a direction AvaloniaUI even wants to start heading down?

**Comment by maxkatz6** [comment]
> I'd love to see greater, low-level support for observables in AvaloniaUI

Avalonia does work with observables in low level. Features like bindings and animations are driven by them internally, with some internal optimizations (like single subscription observables, where it's possible). As a framework user you can find some API and extensions that work with observables too. To bring an example, there are [GetObservable (from a property)](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaObjectExtensions.cs#L39) (and more in AvaloniaObjectExtensions), [GetResourceObservable](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Styling/Controls/ResourceNodeExtensions.cs#L62), [creating binding from an observable](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Data/InstancedBinding.cs#L105).
ReactiveUI also adds some other extensions on top of it.

> how much is the inclusion of observables and ReactiveUI seen as a draw by experienced AvaloniaUI developers

Observables being as part of Avalonia core wasn't a problem for us or anybody else as far as I can tell. @grokys might remember more.
Except of performance issues from generic System.Reactive implementations, but as I said above, Avalonia has internal light-weight implementations for their interfaces, which solves a problem mostly.

ReactiveUI is another story. Just for case, reactive extensions (System.Reactive) is not the same as ReactiveUI. The later one is MVVM framework build on top of reactive extensions with its own sugars. And Avalonia does not use it internally, but provides integration package ([Avalonia.ReactiveUI](https://www.nuget.org/packages/Avalonia.ReactiveUI/)). So, it's completely optional for developers.
ReactiveUI has its own problem of performance, it's fine for almost everything, but slows down for complex applications with thousands of   reactive objects. Also, it has higher complexity for beginners, who weren't familiar with reactive programming before. 
That discussion you attached is not really a problem, because Splat is not required to use even if you are using ReactiveUI on your project. Yes, ReactiveUI does still use it internally AFAIK, but as a developer you are [free to use](https://github.com/AvaloniaUI/Avalonia/discussions/6157#discussioncomment-943957) any other dependency container.

> observables are a core part of Angular; is that a direction AvaloniaUI even wants to start heading down

Yes, reactive extensions won't be removed from Avalonia. 
And no, it does not mean, that Avalonia requires everybody to use ReactiveUI. I would say more, Avalonia.ReactiveUI integration package might be removed from Avalonia org and became independed or part of ReactiveUI org (something like ReactiveUI.Avalonia), although no quaranties here.

**Reply by BenMakesGames** [reply]
ooh, thanks for pointing out those observable extensions; I'll check them out!

re: Splat: I got some bootstrap code working using MS's DI, which is great, although I'm actually finding I don't know exactly how I want to use DI within the UI. obviously, services can depend on one another... but how the views & viewmodels get those services feels a little less certain to me.

I know ReactiveUI and System.Reactive aren't the same... it's ReactiveUI, though, that makes hooking up change notification way easier, via `[Reactive]`. I guess I assumed that Avalonia was leveraging ReactiveUI to do the XAML observable binding. I guess I assumed wrong :P

thanks again for all the info! I'm still definitely learning/wrapping my head around the wild world of WPF/Avalonia/XAML/MVVM

**Reply by BenMakesGames** [reply]
is there documentation about these classes and how they're used/how to use them? (ResourceProvider, AvaloniaObject, etc), or is the best way to learn to look at the code?

**Reply by maxkatz6** [reply]
I don't think there is a doc page about these specific properties.
